File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	OpcodeData byte = 0
	OpcodeAck  byte = 1
	OpcodeErr  byte = 2
	OpcodeFin  byte = 3
)

const opcodeSize = 1
const lengthSize = 4
const headerSize = opcodeSize + lengthSize
const maxPayloadSize = (uint64(1) << (lengthSize * 8)) - 1 // 2^32 bits max

type Protocol struct {
	socket io.ReadWriter
	buffer []byte
}

func NewProtocol(socket io.ReadWriter) *Protocol {
	return &Protocol{
		socket: socket,
		buffer: make([]byte, 0),
	}
}

func (protocol *Protocol) getBytes(size int) ([]byte, error) {
	for len(protocol.buffer) < size {
		missing := size - len(protocol.buffer)
		chunk, err := safe_socket.RecvAll(protocol.socket, missing)
		if err != nil {
			return nil, err
		}
		protocol.buffer = append(protocol.buffer, chunk...)
	}

	result := append([]byte(nil), protocol.buffer[:size]...)
	protocol.buffer = protocol.buffer[size:]
	return result, nil
}

func (protocol *Protocol) RecvMessage() (byte, []byte, error) {
	header, err := protocol.getBytes(headerSize)
	if err != nil {
		return 0, nil, err
	}

	opcode := header[0]
	payloadLength := int(binary.BigEndian.Uint32(header[opcodeSize:headerSize]))
	payload, err := protocol.getBytes(payloadLength)
	if err != nil {
		return 0, nil, err
	}

	return opcode, payload, nil
}

func (protocol *Protocol) createMessage(opcode byte, payload []byte) ([]byte, error) {
	if uint64(len(payload)) > maxPayloadSize {
		return nil, fmt.Errorf("payload size %d exceeds maximum %d", len(payload), maxPayloadSize)
	}

	message := make([]byte, headerSize+len(payload))
	message[0] = opcode
	binary.BigEndian.PutUint32(message[opcodeSize:headerSize], uint32(len(payload)))
	copy(message[headerSize:], payload)
	return message, nil
}

func (protocol *Protocol) SendMessage(opcode byte, payload []byte) error {
	message, err := protocol.createMessage(opcode, payload)
	if err != nil {
		return err
	}
	return safe_socket.SendAll(protocol.socket, message)
}

func (protocol *Protocol) SendDataMessage(payload []byte) error {
	return protocol.SendMessage(OpcodeData, payload)
}

func (protocol *Protocol) SendAck() error {
	return protocol.SendMessage(OpcodeAck, []byte{})
}

func (protocol *Protocol) SendErr() error {
	return protocol.SendMessage(OpcodeErr, []byte{})
}

func (protocol *Protocol) SendFin() error {
	return protocol.SendMessage(OpcodeFin, []byte{})
}

File: services/server/src/protocol/protocol.py
import socket
import safe_socket

OPCODE_DATA = 0
OPCODE_ACK = 1
OPCODE_ERR = 2
OPCODE_FIN = 3

_OPCODE_SIZE = 1
_LENGTH_SIZE = 4
_HEADER_SIZE = _OPCODE_SIZE + _LENGTH_SIZE
_MAX_PAYLOAD_SIZE = (1 << (_LENGTH_SIZE * 8)) - 1  # 2^32 bits max

class Protocol:
	def __init__(self, socket: socket.socket):
		self._socket = socket
		self._buffer = bytearray()

	def _get_bytes(self, size: int):
		while len(self._buffer) < size:
			missing = size - len(self._buffer)
			chunk = safe_socket.recv_all(self._socket, missing)
			self._buffer.extend(chunk)

		result = self._buffer[:size]
		del self._buffer[:size]
		return bytes(result)

	def recv_message(self) -> tuple[int, bytes]:
		header = self._get_bytes(_HEADER_SIZE)
		opcode = header[0]
		payload_length = int.from_bytes(header[_OPCODE_SIZE:_HEADER_SIZE], byteorder='big')
		return opcode, self._get_bytes(payload_length)

	def _create_message(self, opcode:int, payload: bytes) -> bytes:
		payload_length = len(payload)
		if payload_length > _MAX_PAYLOAD_SIZE:
			raise ValueError(
				f"payload size {payload_length} exceeds maximum {_MAX_PAYLOAD_SIZE}"
			)
		opcode_bytes = opcode.to_bytes(_OPCODE_SIZE, byteorder="big")
		length_bytes = payload_length.to_bytes(_LENGTH_SIZE, byteorder='big')
		return opcode_bytes + length_bytes + payload

	def send_message(self, opcode: int, payload: bytes = b""):
		msg = self._create_message(opcode, payload)
		safe_socket.send_all(self._socket, msg)
	
	def send_data(self, payload: bytes):
		self.send_message(OPCODE_DATA, payload)

	def send_ack(self):
		self.send_message(OPCODE_ACK, b"")

	def send_err(self):
		self.send_message(OPCODE_ERR, b"")

	def send_fin(self):
		self.send_message(OPCODE_FIN, b"")